internal/database/migration/store/store.go

```diff
@@ -101,36 +101,93 @@ func (s *Store) EnsureSchemaTable(ctx context.Context) (err error) {
 // table either represents a migration that needs to be applied, or a migration defined in a version
 // prior to the instance's first boot. Backfilling these records prevents the latter circumstance as
 // being interpreted as the former.
+//
+// DO NOT call this method from inside a transaction, otherwise the absence of optional relations
+// will cause a transaction rollback while this function returns a nil-valued error (hard to debug).
 func (s *Store) BackfillSchemaVersions(ctx context.Context) error {
-	// Choose the lowest relevant version (most like the smallest squashed migration) that has
-	// been successfully applied on this instance. We will be backfilling all ancestors of this
-	// migration version given the stitched migration graph.
-	version, ok, err := s.inferBackfillTarget(ctx)
+	stitchedDefinitions := shared.StitchedMigationsBySchemaName[humanizeSchemaName(s.schemaName)].Definitions
+
+	applied, pending, failed, err := s.Versions(ctx)
 	if err != nil {
 		return err
 	}
-	if !ok {
+	if len(pending) != 0 || len(failed) != 0 {
+		// If we have a dirty database here don't overwrite in-progress/failed records with fake
+		// successful ones. This would end up masking a lot of drift conditions that would make
+		// upgrades painful and operation of the instance unstable.
 		return nil
 	}
 
-	// Determine ancestors of the chosen root
-	ancestorDefinitions, err := s.stitchedMigration().Definitions.Up(nil, []int{version})
+	// We have a small fork in behavior coming up. If we read from the migration_logs table, then
+	// we already had a migration log entry for the last version we applied. However, if we read
+	// from the old golang-migrate tables, we have an empty migration_log table and want to populate
+	// the log for the latest version as well. If we read from golang-migrate tables we'll set this
+	// flag and no-op
+	writeAppliedVersions := false
+
+	if len(applied) == 0 {
+		// Fallback to golang migrate, but only if there's no authoritative data. This reads
+		// the old .*schema_migrations table (if it exists) and sets the version number as the
+		// sole applied (faux) migration log record. This number still corresponds correctly to
+		// a node in the stitched migration graph.
+
+		version, ok, err := basestore.ScanFirstInt(s.Query(ctx, sqlf.Sprintf(inferBackfillTargetViaGolangMigrateQuery, quote(s.schemaName))))
+		if err != nil && !isMissingRelation(err) {
+			return err
+		}
+		if !ok {
+			return nil
+		}
+
+		writeAppliedVersions = true
+		applied = append(applied, version)
+	}
+
+	// Make lookup map of applied migration identifiers
+	appliedMap := make(map[int]struct{}, len(applied))
+	if !writeAppliedVersions {
+		for _, id := range applied {
+			appliedMap[id] = struct{}{}
+		}
+	}
```

Comment:
Please correct if wrong:

```suggestion
	// Later on, members of appliedMap that are ancestors of applied
	// will be backfilled into the database. To avoid writing them,
	// we do not populate appliedMap
	if !writeAppliedVersions {
		for _, id := range applied {
			appliedMap[id] = struct{}{}
		}
	}
```

Reply:
```suggestion
	// Presence of an identifier in appliedMap will prevent a migration log for
	// that definition to be omitted from the writes. In order to ensure that we
	// write the applied versions (in the case of golang-migrate), we'll leave
	// this map empty.
	if !writeAppliedVersions {
		for _, id := range applied {
			appliedMap[id] = struct{}{}
		}
	}
```

This makes sense?

Reply:
actually now that I'm reading that comment I'm definitely getting it wrong - if we _don't want to write applied versions_, we _do want to populate appliedMap_

Reply:
yes thank you!